File: 프로그래머스/KaKao/오픈채팅방.swift
// 2019 KAKAO BLIND RECRUITMENT 오픈채팅방
// https://school.programmers.co.kr/learn/courses/30/lessons/42888
import Foundation

func solution(_ record:[String]) -> [String] {
    var answer:[String] = []
    let idNickDict: [String: String] = makeRelationshipDictionary(record)
    
    for rec in record {
        let splitedR = rec.components(separatedBy: " ")
        if splitedR[0] == "Enter" {
            answer.append("\(idNickDict[splitedR[1]]!)님이 들어왔습니다.")
        } else if splitedR[0] == "Leave" {
            answer.append("\(idNickDict[splitedR[1]]!)님이 나갔습니다.")
        }
    }
    return answer
}

func makeRelationshipDictionary(_ record:[String]) -> [String: String] {
    var ret: [String: String] = [:]
    
    for r in record {
        let splitedR = r.components(separatedBy: " ")
        if splitedR[0] == "Leave" { continue }
        ret[splitedR[1]] = splitedR[2] 
    }
    return ret
}

File: 프로그래머스/KaKao/오픈채팅방.py
# 2019 KAKAO BLIND RECRUITMENT 오픈채팅방
def solution(record):
    answer = []
    stateDic = {}

    for r in record:
        tmp = r.split(" ")
        # print(tmp)

        if len(tmp) > 2:
            stateDic[tmp[1]] = tmp[2]

    for r in record:
        tmp = r.split(" ")

        if tmp[0] == "Enter":
            answer.append(stateDic[tmp[1]]+"님이 들어왔습니다.")
        elif tmp[0] == "Leave":
            answer.append(stateDic[tmp[1]]+"님이 나갔습니다.")

    return answer
